### PR TITLE
Refactor documentation

### PR DESCRIPTION
**Pull request by xkubov**
This MR contains large refactor of IS Core documentation providing detailed overview, architecture and thorough examples on how to use the library.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,8 +33,34 @@ Docs
 For more details see [docs](http://django-is-core.readthedocs.org/)
 
 
-Contribution
-------------
+Documentation
```
Comment (xkubov) [inline README.md]: We should add a section on how to install virtulenv before this. This section depends on activating virtualenv from the `example` app

File: README.md [hunk 2]
```diff
@@ -33,8 +33,34 @@ Docs
...
+   python -m http.server 8000 --directory .build/html
+   ```
+   Then visit http://localhost:8000
+
+### Live Reload (for Contributors)
 
 To run Livereload for Sphinx you need [livereload](https://pypi.python.org/pypi/livereload) `pip install livereload`.
-After installing simply call `make htmllivereload` and open [http://localhost:5500/](http://localhost:5500/).
\ No newline at end of file
+After installing simply call `make htmllivereload` and open [http://localhost:5500/](http://localhost:5500/).
```
Comment (xkubov) [inline README.md]: this step doesn't work - we should either make it work or discard this section

File: docs/architecture.rst [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+   :alt: IS Core Architecture Diagram
+   :align: center
+   :width: 200%
+
+The flow in detail:
+
+1. **Initial Request**: Browser hits a URL → View returns HTML template with links to JS/static files
+2. **Static Assets**: Browser loads JavaScript, CSS, and images from static files
+3. **Data Loading**: JavaScript calls the appropriate Core's REST endpoint for data (e.g., to populate a table)
+4. **Interactive Updates**: Further interactions (filtering, row actions) go through REST → JS updates the DOM
```
Comment (xkubov) [inline docs/architecture.rst:40]: This should be updated so that it is clear that this is not typical request but example request that showcases  server-side and client-side rendering flow.

File: docs/architecture.rst [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+Types of Core
+-------------
+
+Django IS Core provides several types of Core classes that can be combined based on your needs:
```
Comment (xkubov) [inline docs/architecture.rst]: Add a comment that cores are described in cores section. Where cores are described in more detail

File: docs/architecture.rst [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+Default REST API over Django models with automatic CRUD endpoints.
+
+Combined Core (DjangoUiRestCore)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The most common case - combines UI + REST in a single Core. Provides both the admin interface and the REST API that powers it.
+
+.. tip::
+   **Recommended**: Use ``DjangoUiRestCore`` for most use cases. It gives you both admin UI and REST API with a single class definition.
```
Comment (xkubov) [inline docs/architecture.rst:100]: can we add links to the references api?

File: docs/architecture.rst [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+Example::
+
+    from is_core.main import DjangoUiRestCore
+
+    class ArticleCore(DjangoUiRestCore):
+        model = Article
+        list_fields = ('title', 'author', 'created_at')
+        form_fields = ('title', 'content', 'author')
```
Comment (xkubov) [inline docs/architecture.rst]: `form_fields` should be ideally just `fields`

File: docs/architecture.rst [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+Registration and Auto-discovery
+--------------------------------
+
+Django IS Core uses automatic registration through metaclasses and singletons:
```
Comment (xkubov) [inline docs/architecture.rst:114]: Specify, that the only thing that library users need to do is create `cores.py` in their application registred in `INSTALLED_APPS`.

File: docs/architecture.rst [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+        # This Core is automatically registered when the module is imported
+        # The CoreBase metaclass calls CoresLoader.register_core(ArticleCore)
+
+Resources and Actions
+---------------------
+
+Generic Resource
+^^^^^^^^^^^^^^^^
+
+A custom endpoint where you implement get/post/put/delete methods and register it via ``get_rest_patterns`` on a specific path.
```
Comment (xkubov) [inline docs/architecture.rst]: can we also describe where this resource needs to be registred? in core? Or even better add link to the section that describes it in more detail.

File: docs/cores.rst [hunk 9]
```diff
@@ -713,27 +848,78 @@ Options
...
+
+        @short_description(_('Coin Balance'))
+        def coin_balance(self, obj):
+            return CoinTransaction.objects.compute_balance_for_customer(obj)
+
+        # For filterable extra fields, use @filter_class decorator
+        @filter_class(AppliedPromocodeIDFilter)
+        @property
+        def applied_promocodes_id(self):
+            return None  # Return value not used, filter handles the logic
```
Comment (xkubov) [inline docs/cores.rst]: not returning any value means `rest_extra_fields` will always return `None` for this property. We should specify that.

File: docs/cores.rst [hunk 10]
```diff
@@ -713,27 +848,78 @@ Options
...
+
+This pattern is useful for:
+
+- **Performance**: Expensive computed fields are only calculated when needed
+- **API flexibility**: Frontend can request additional data without changing default payload
+- **Separation**: REST-only data doesn't clutter UI field definitions
```
Comment (xkubov) [inline docs/cores.rst]: Remove this section.

File: docs/cores.rst [hunk 11]
```diff
@@ -774,3 +955,281 @@ The atribute inside response has named ``_class_names``.
...
+                kwargs["customer_pk"] = (
+                    request.kwargs.get("customer_pk")
+                    if hasattr(request, "kwargs")
+                    else None
+                )
+            return kwargs
+
+
+    class CustomerCommentRestPattern(CustomerCommentBasePattern, RestPattern):
+        pass
```
Comment (xkubov) [inline docs/cores.rst]: remove this section

File: docs/cores.rst [hunk 12]
```diff
@@ -774,3 +955,281 @@ The atribute inside response has named ``_class_names``.
...
+**Real-World Example:**
+
+In the example_customers application, CustomerCore doesn't define device management at all. Instead:
+
+- ``CustomerMobileDeviceCore`` in ``cores/customer/device.py`` handles all device logic
+- ``CustomerLimitChangeCore`` in ``cores/customer/limit_change.py`` handles limit changes
+- ``CustomerCommentCore`` in ``cores/customer/comment.py`` handles comments
+
+Each gets full UI, REST API, export, and independent evolution—all while maintaining the nested URL structure through custom patterns.
+
```
Comment (xkubov) [inline docs/cores.rst]: remove this section

File: docs/filters.rst [hunk 13]
```diff
@@ -1,35 +1,593 @@
...
+Example::
+
+    class ArticleCore(DjangoUiRestCore):
+        model = Article
+        list_fields = ('title', 'word_count')
+
+        def word_count(self, obj):
+            return len(obj.content.split())
+        word_count.filter = True  # Enable filtering
+        word_count.filter_by = 'content'  # Filter by content field
```
Comment (xkubov) [inline docs/filters.rst]: Remove this section

File: docs/filters.rst [hunk 14]
```diff
@@ -1,35 +1,593 @@
...
+        model = Order
+
+        def get_list_filter(self, request):
+            manager = super().get_list_filter(request)
+            manager.add_filter(
+                'payment_method_type',
+                PaymentMethodFilter,
+                enum_class=PaymentMethodEnum
+            )
+            return manager
```
Comment (xkubov) [inline docs/filters.rst]: remove this section.

File: docs/forms.rst [hunk 15]
```diff
@@ -1,3 +1,674 @@
...
+                'accept': 'image/*'
+            })
+        )
+
+**Features:**
+
+- Thumbnail preview of current image
+- "Clear" checkbox to remove existing file
+- File size display
+- Upload progress indication
```
Comment (xkubov) [inline docs/forms.rst]: Remove this section.

File: docs/forms.rst [hunk 16]
```diff
@@ -1,3 +1,674 @@
...
+
+3. **Use fieldsets** to organize complex forms into logical sections
+
+4. **Implement proper validation** at the appropriate level (field, form, or model)
+
+5. **Make forms dynamic** when needed but keep the logic clear and maintainable
+
+6. **Use custom widgets** for specialized input types (date pickers, file uploads, etc.)
+
+7. **Test form validation** thoroughly, especially for complex multi-field validation
```
Comment (xkubov) [inline docs/forms.rst]: Remove this section.

File: docs/forms.rst [hunk 17]
```diff
@@ -1,3 +1,674 @@
...
+        fields = ('first_name', 'last_name', 'email')
+
+        # Append after parent's fieldsets
+        fieldsets_postfix = CommentCoreMixin.notes_fieldset + (
+            (_l('Bank Accounts'), {
+                'inline_view': CustomerBankAccountInlineTableView,
+            }),
+        )
+
+This preserves the parent's fieldset configuration while adding your own.
```
Comment (xkubov) [inline docs/forms.rst:590]: Remove this section.

File: docs/menu.rst [hunk 18]
```diff
@@ -1,12 +1,605 @@
...
+                settings = MenuGroup('settings', 'Settings')
+                settings.add_item(MenuItem(
+                    'users',
+                    'Users',
+                    url=reverse('admin:user-list'),
+                    icon='users'
+                ))
+                menu.append(settings)
+
+            return menu
```
Comment (xkubov) [inline docs/menu.rst]: Remove this section

File: docs/menu.rst [hunk 19]
```diff
@@ -1,12 +1,605 @@
...
+            # Add custom settings at the end
+            if request.user.is_superuser:
+                menu.append(MenuItem(
+                    'settings',
+                    'Settings',
+                    url='/admin/settings/',
+                    icon='settings'
+                ))
+
+            return menu
```
Comment (xkubov) [inline docs/menu.rst]: Remove this section

File: docs/menu.rst [hunk 20]
```diff
@@ -1,12 +1,605 @@
...
+
+6. **Use icons consistently**: Pick one icon set and stick with it
+
+7. **Show context**: Use badges to display counts or notifications
+
+8. **Test with different roles**: Verify menu appears correctly for all user types
+
+9. **Consider mobile**: Ensure menu works well on small screens
+
+10. **Document custom menus**: If using a custom generator, document the structure for other developers
```
Comment (xkubov) [inline docs/menu.rst]: Remove this section.

File: docs/rests.rst [hunk 21]
```diff
@@ -1,7 +1,996 @@
+.. _rests:
 
-REST
-====
+REST Resources
+==============
+
+Django IS Core provides a powerful REST API system built on top of django-pyston. Resources define REST endpoints that serve as the backend for UI actions and can be used independently to build custom frontends.
```
Comment (xkubov) [inline docs/rests.rst]: Tone down the marketing. Be more pragmatic, narrate what it is, don't try to sell how "powerful" it is.

File: docs/rests.rst [hunk 22]
```diff
@@ -1,7 +1,996 @@
...
+
+6. **Use appropriate HTTP methods** - GET for reads, POST for creates, PUT/PATCH for updates, DELETE for deletes
+
+7. **Return proper status codes** - 200 for success, 201 for created, 400 for validation errors, 403 for permission denied, 404 for not found
+
+8. **Document custom endpoints** so frontend developers know how to use them
+
+9. **Version your API** if you plan to make breaking changes
+
+10. **Test REST endpoints** thoroughly, including error cases and edge conditions
```
Comment (xkubov) [inline docs/rests.rst]: Remove this section.

File: docs/architecture.rst [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+                    'url': self.get_edit_url(obj),
+                    'type': 'ui'
+                })
+            if request.user.has_perm('articles.delete_article'):
+                actions.append({
+                    'name': 'delete',
+                    'url': self.get_api_url('delete', obj.pk),
+                    'type': 'api'
+                })
+            return actions
```
Comment (xkubov) [inline docs/architecture.rst:200]: add references to the place where it is more thoroughly described.

**Comment by xkubov** [comment]
image

File: README.md [hunk 24]
```diff
@@ -33,8 +33,88 @@ Docs
...
+
+2. Install Sphinx and the ReadTheDocs theme:
+   ```bash
+   pip install sphinx sphinx_rtd_theme
+   ```
+
+3. Build the HTML documentation:
+   ```bash
+   cd docs
+   make html
```
Comment (vichi99) [inline README.md]: It doesnt work.

```
uv venv --python 3.11 
uv init
uv add sphinx sphinx_rtd_theme
source .venv/bin/activate
cd docs
make html
```

ModuleNotFoundError: No module named 'django'

I had to install 'django'. Could you add it to the README, including the version?


Then I was able to generate the docs, it works on `python -m http.server 8000 --directory .build/html`, ok.
However, I got these warnings. Could you please fix or describe this in the README? 

Image

File: README.md [hunk 25]
```diff
@@ -33,8 +33,88 @@ Docs
...
+### Quick Start
+
+1. Navigate to the example directory:
+   ```bash
+   cd example
+   ```
+
+2. Install and set up the application:
+   ```bash
+   make install
```
Comment (vichi99) [inline README.md:61]: After `make install` I got.

```
find . -name "*.pyc" -delete;
find . -type d -empty -delete;
rm -rf var/ve
rm -rf /Users/jan.vicha/Projects/work/skippay/django-is-core/example/var
virtualenv -p python3 var/ve
bash: virtualenv: command not found
make: *** [initvirtualenv] Error 127
```

File: docs/views.rst [hunk 26]
```diff
@@ -1,60 +1,848 @@
...
+
+Generic form views for various use cases.
 
 BulkChangeFormView
+^^^^^^^^^^^^^^^^^^
+
+Allows bulk editing of multiple objects at once.
+
+**Use case:** Change status of multiple articles from "draft" to "published"
+
```
Comment (vichi99) [inline docs/views.rst:248]: - There is no example to use BulkChangeFormView. However, we didnt use it in Core, so I guess its fine.
- It is not mentioned that bulk_change_enabled is set to True.
- bulk_change_fields is missing.

File: docs/menu.rst
Comment (vichi99) [inline docs/menu.rst:1]: - Missing a description for CollapsibleMenuItem and an example. We use this as a BackendGroupMenuItem in Core.
- nitpick: Missing a description for MenuGenerator and an example. There is only for CustomMenuGenerator, but maybe that is purpose.

File: docs/cores.rst
Comment (vichi99) [inline docs/cores.rst:1]: It would be good to add an explanation for Mixin Composition.
- The order of mixins (Method Resolution Order).
- How mixins work together.
- Best practices.

**Review by vichi99 — COMMENTED** [review]
Good job overall, just a few minor things to fix.

File: README.md (hunk 24 above)
Comment (xkubov) [reply README.md]: Hm, I admit the instructions are not clear. I'll clear that up. 

btw. you need to use venv from the `example` app. By using that you'll have the required django dep.

File: README.md (hunk 25 above)
Comment (xkubov) [reply README.md:61]: I'll take a look at those instructions

File: README.md (hunk 24 above)
Comment (xkubov) [reply README.md]: It should be clear how to setup the project now. Can you try?

File: docs/cores.rst
Comment (xkubov) [reply docs/cores.rst:1]: We can in separate MR. I think mixin part is nice to have. Purpose of this MR was primarily to add instructions on how to run the project and add architecture description.

File: docs/views.rst (hunk 26 above)
Comment (xkubov) [reply docs/views.rst:248]: added

File: docs/menu.rst
Comment (xkubov) [reply docs/menu.rst:1]: I think it is not required. `CollapsibleMenuItem` is redundant class. It just initializes MenuItem with `collapsible=True`.  I would much rather not have it in is core at all.

File: README.md (hunk 24 above)
Comment (vichi99) [reply README.md]: It is clearer now and it works. However, there are still warnings when building the documentation. Should we leave them like this?

image

File: README.md (hunk 25 above)
Comment (vichi99) [reply README.md:61]: It is working now.

File: docs/views.rst (hunk 26 above)
Comment (vichi99) [reply docs/views.rst:248]: checked

File: docs/menu.rst
Comment (vichi99) [reply docs/menu.rst:1]: Okay, we can discuss this in another task.

File: docs/cores.rst
Comment (vichi99) [reply docs/cores.rst:1]: Okay, we can discuss this in another task.

File: README.md (hunk 24 above)
Comment (xkubov) [reply README.md]: I fixed those warnings.

File: docs/reference/resources.rst [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+   :show-inheritance:
+
+Django Core Resources
+---------------------
+
+.. autoclass:: is_core.rest.resource.DjangoCoreResource
+   :members:
+   :show-inheritance:
+
+.. autoclass:: is_core.rest.resource.DjangoResource
```
Comment (vichi99) [inline docs/reference/resources.rst]: References to is_core.rest.resource.BaseResource and is_core.rest.resource.DjangoResource are incorrect - these classes are imported from pyston.resource, not defined in is_core.rest.resource.

Comment (xkubov) [reply docs/reference/resources.rst]: fixed.

**Review by vichi99 — APPROVED** [review]
